src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Utilities/StreamExtensions.cs

```diff
@@ -0,0 +1,123 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.ProjectSystem;
+
+#if !NET
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+#endif
+
+namespace Microsoft.AspNetCore.Razor.Utilities;
+
+internal static class StreamExtensions
+{
+    public static Task WriteStringAsync(this Stream stream, string text, Encoding? encoding = null, CancellationToken cancellationToken = default)
+    {
+        Debug.Assert(stream.CanWrite);
+        encoding ??= Encoding.UTF8;
+
+        var encodedBytes = encoding.GetBytes(text);
```

Comment:
How often is this method called? This is allocating `byte[]` for every `string` that is passed in. If this is done with any frequency consider using an array pool and overrides of `GetBytes` that take a `Span<byte>`. May have to polyfill a bit on that.

Reply:
Good point, I converted a lot of things to array pool but missed this one!